lang/cs/Org.Apache.REEF.Wake/Remote/Impl/DefaultRemoteManager.cs

```diff
@@ -142,22 +147,35 @@ public IObserver<T> GetRemoteObserver(IPEndPoint remoteEndpoint)
             }
 
             ProxyObserver remoteObserver;
-            if (!_cachedClients.TryGetValue(remoteEndpoint, out remoteObserver))
+            if (!_enableCaching)
             {
-                TransportClient<IRemoteEvent<T>> client = 
-                    new TransportClient<IRemoteEvent<T>>(remoteEndpoint, _codec, _observerContainer);
-                var msg = string.Format("NewClientConnection: Local {0} connected to Remote {1}",
-                    client.Link.LocalEndpoint.ToString(), 
-                    client.Link.RemoteEndpoint.ToString());
-                LOGGER.Log(Level.Info, msg);
-
-                remoteObserver = new ProxyObserver(client);
+                remoteObserver = CreateRemoteObserver(remoteEndpoint);
+            }
+            else if (!_cachedClients.TryGetValue(remoteEndpoint, out remoteObserver))
+            {
+                remoteObserver = CreateRemoteObserver(remoteEndpoint);
                 _cachedClients[remoteEndpoint] = remoteObserver;
             }
 
             return remoteObserver;
         }
 
+        private ProxyObserver CreateRemoteObserver(IPEndPoint remoteEndpoint)
```

Comment:
@afchung  don't see any other option, besides returning our own object instead IObserver, that will implement both IObserver and IDisposable. I don't think that in general caching can work well here, and should  be done on client side. WDYT?